lib/livebook/session.ex

```diff
@@ -157,8 +157,21 @@ defmodule Livebook.Session do
   """
   @spec fetch_assets(pid(), String.t()) :: :ok | {:error, String.t()}
   def fetch_assets(pid, hash) do
-    # TODO make this async using the worker-registry pattern
-    GenServer.call(pid, {:fetch_assets, hash})
+    case GenServer.call(pid, {:handle_fetch_assets, hash}) do
```

Comment:
This is still centralized through the session. You want to inline the code inside `GenServer.call` here. You will only reach out to the session process inside the task and only to ask it where the runtime is located.

Reply:
We need the session to also determine the archive path, but we can move the initial check to the caller yeah, just did that :D Also if there is no runtime or archive we just return an error without starting the task.